downstairs/src/extent_inner_raw_v2.rs

```diff
@@ -0,0 +1,1058 @@
+    fn flush_number(&self) -> Result<u64, CrucibleError> {
+        self.get_metadata().map(|v| v.flush_number)
+    }
+
+    fn gen_number(&self) -> Result<u64, CrucibleError> {
+        self.get_metadata().map(|v| v.gen_number)
+    }
+
+    fn dirty(&self) -> Result<bool, CrucibleError> {
+        Ok(self.dirty)
+    }
+
+    fn write(
+        &mut self,
+        job_id: JobId,
+        write: &ExtentWrite,
+        only_write_unwritten: bool,
+        iov_max: usize,
+    ) -> Result<(), CrucibleError> {
+        check_input(self.extent_size, write.offset, write.data.len())?;
+        let n_blocks = write.block_contexts.len();
+
+        let start_block = write.offset;
+        let block_size = self.extent_size.block_size_in_bytes();
+        let mut block = start_block;
+
+        struct WriteChunk<'a> {
+            offset: BlockOffset,
+            iovecs: Vec<IoSlice<'a>>,
+        }
+        let mut iovecs = vec![];
+
+        let ctxs = write
+            .block_contexts
+            .iter()
+            .map(|ctx| {
+                let mut buf = [0u8; BLOCK_CONTEXT_SIZE_BYTES as usize];
+                let ctx = ReadBlockContext::from(*ctx);
+                bincode::serialize_into(buf.as_mut_slice(), &ctx).unwrap();
+                buf
+            })
+            .collect::<Vec<_>>();
+
+        let padding = vec![0u8; self.layout.padding_size() as usize];
+        let mut expected_bytes = 0;
+        let mut start_new_chunk = true;
+        for (data, ctx) in write.data.chunks(block_size as usize).zip(&ctxs) {
+            if only_write_unwritten && self.block_written[block.0 as usize] {
+                start_new_chunk = true;
+                block.0 += 1;
+                continue;
+            }
+            if start_new_chunk {
+                iovecs.push(WriteChunk {
+                    offset: block,
+                    iovecs: vec![],
+                });
+                start_new_chunk = false
+            }
+            let vs = &mut iovecs.last_mut().unwrap().iovecs;
+            vs.push(IoSlice::new(data));
+            vs.push(IoSlice::new(ctx));
+            expected_bytes += data.len() + ctx.len();
+            if self.layout.has_padding_after(block) {
+                vs.push(IoSlice::new(&padding));
+                expected_bytes += padding.len();
+            }
+            block.0 += 1;
+
+            // If the next write would exceed iov_max, set the `start_new_chunk`
+            // flag so that we begin a new chunk before pushing the write.
+            start_new_chunk |=
+                vs.len() + 2 + self.layout.has_padding_after(block) as usize
+                    > iov_max;
+        }
+
+        if iovecs.is_empty() {
+            return Ok(());
+        }
+
+        self.set_dirty()?;
+
+        cdt::extent__write__file__start!(|| {
+            (job_id.0, self.extent_number.0, n_blocks as u64)
+        });
+
+        // Now execute each chunk in a separate `pwritev` call
```

Comment:
TODO-jwm this means the Crucible level write can partially succeed

Reply:
To be clear! I think this was always true but was TODO for myself to 1) verify this and 2) suggest we add this as documentation somewhere?